### ISSUE TITLE
Strange "Undefined binding(s) detected" error message

### ISSUE THREAD

**Issue by leventov** [knex/knex#3723]
## Environment

Knex version: 0.20.10
Database + version: Postgres 12

## Bug

1. Explain what kind of behaviour you are getting and how you think it should do

```
knex.raw("?", [
  knex.insert({ id: "b", fields: { a: undefined } }).into("test"),
])
```

2. Error message

```Error: Undefined binding(s) detected for keys [] when compiling RAW query: insert into "organizations" ("fields", "id") values (?, ?)```

This error message is puzzling. Undefined keys are not printed.

**Comment by kibertoad** [comment]
This is a valid complaint. We already improved this for empty updates, and I don't see why not do same here.

**Comment by thekuom** [comment]
@kibertoad are you working on this one?

**Comment by kibertoad** [comment]
@thekuom Thanks for the ping, gonna add it now

**Comment by thekuom** [comment]
@kibertoad Oh I was wondering if I could start on it as long as you hadn't already done it 😅

**Comment by kibertoad** [comment]
@thekuom It's less straightforward than I though, due to rather cluttered logic around this place, and a peculiar way raw insert is being done in original case. If you have some time to dig deeper, here's a test case you can use for reproducing the issue:

```
const Knex = require('../../../lib/index');
const { expect } = require('chai');

describe('querybuilder', () => {
  let knex;
  beforeEach(() => {
      knex = Knex({
        client: 'sqlite',
        connection: ':memory:',
      });
    },
  );

  it('should throw an informative error on undefined binding', async () => {
    try {
      await knex.raw('?', [
        knex.insert({ id: 'b', fields: { a: undefined } }).into('test'),
      ]);
    } catch (err) {
      expect(err.message).eql('Undefined binding(s) detected for keys [0] when compiling RAW query: insert into `test` (`fields`, `id`) values (?, ?)');
    }
  });
});
```

@leventov For curiosity sake, why you building query in a such complex way?

Both of these should already work properly:

```
knex.raw('?', [undefined])
```
or 
```
knex.raw(':field', { field: undefined })
```

Does your query work correctly when value is not undefined?

**Comment by thekuom** [comment]
I have some time. I'll let you know if I have any questions

**Comment by thekuom** [comment]
@kibertoad I actually think the insert should succeed and the error should not be thrown. My understanding is that "fields" is a json column on "test"

**Comment by kibertoad** [comment]
@thekuom Inserting undefined values is not allowed. If intent is to insert null, it should be passed as null.

**Comment by kibertoad** [comment]
@thekuom And what I meant in my original comment is that there is a weird mix between raw and non-raw querying in that example. Second parameter for raw query is for passing bindings for placeholder in the raw query, but instead of doing that it passes non-raw select as a binding, so the whole intent of this is very puzzling for me. If idea is to insert `{ id: 'b', fields: { a: undefined } }` into `test`, there is no reason to wrap that into knex.raw , it would work perfectly fine as `knex('test').insert({ id: 'b', fields: { a: undefined } }`)`

**Comment by thekuom** [comment]
@kibertoad I am confused about that as well, since `knex.insert({ id: 'b', fields: { a: undefined } }).into('test')` works as I would expect assuming "fields" is a json column. It inserts:
```
| id  | fields |
| 'b' | {}        |
```

**Comment by kibertoad** [comment]
Yes, problem is with knex.raw specifically. @leventov Can you explain what knex.raw is for there?

**Comment by leventov** [comment]
I think I encountered this problem here: https://github.com/climatescape/climatescape.org/blob/cbf63e76c80317150ce9ce67545419719d22d3a3/backend/src/db/pg.js#L124-L146

**Comment by thekuom** [comment]
@kibertoad, after digging around, it appears that `obj.bindings` [here](https://github.com/knex/knex/blob/master/lib/raw.js#L106) is `'b', { a: undefined }`, which will pass the condition and eventually get [here](https://github.com/knex/knex/blob/master/lib/helpers.js#L31). Because the binding in this case is a `knex.insert` statement, `toSQL` is a function and the responsibility for determining the undefined indices is delegated to that statement.

So the problem (IMO) boils down to: 
- `knex.raw` is giving a false positive on undefined bindings because `knex.insert.toSQL` is seemingly fine with `knex.insert({ id: 'b', fields: { a: undefined })` OR
- `knex.raw` is correct that there are illegal undefined bindings and `knex.insert.toSQL` is not properly throwing an error for it

You mentioned that inserts with undefined should be illegal, but current functionality is that they work at least for this case (without the raw). So this ends up being a decision whether to pass this case, or fail existing cases. What are your thoughts?

**Comment by genadis** [comment]
I believe this https://github.com/knex/knex/issues/3548 is related

**Comment by mercmobily** [comment]
This needs an error‑message improvement. The current “Undefined binding(s) detected for keys [] when compiling RAW query: …” does not show the offending binding index in your case. It should surface which binding is undefined (for example `[0]`).

Code context (from the thread): for `knex.raw('?', [ knex('test').insert({...}) ])`, `obj.bindings` in `lib/raw.js` (around the spot referenced at the time, line 106) becomes `['b', { a: undefined }]`, which passes the basic check and then flows into the helpers path (see `lib/helpers.js` around the undefined‑binding handling). Because the inner insert builder’s `toSQL()` is delegated to from inside `raw`, the information about which outer binding index was undefined is lost and the error reports `[]`. The end result is an unhelpful message even though there is a genuine undefined binding.

Inserting `undefined` should be treated as an error; if you intend `NULL`, pass `null` explicitly. The error message should make that obvious and point to the index.

kibertoad's sketch reproduces the issue precisely.

I think we should:
- Improve the undefined binding error to include indices for nested structures, matching the test expectation above (`[0]`). Concretely, track the outer binding indices while delegating to nested builders so that the error assembly in helpers can report them.
- Maybe consider adding a lint to discourage wrapping an `insert` in `knex.raw('?', [...])` when unnecessary.

References from the discussion:
- Quote: “`obj.bindings` here is `'b', { a: undefined }`, which will pass the condition and eventually get here …” -- pointing at `lib/raw.js` and `lib/helpers.js` as the relevant path.
- Real‑world case: climatescape.org code path (link in thread) triggers the same unhelpful message.

Related issues: #3548 (similar binding confusion)
